tools/slicec-cs/src/generators/exception_generator.rs

```diff
@@ -44,118 +56,144 @@ pub fn generate_exception(exception_def: &Exception) -> CodeBlock {
-    let exception_name = exception_def.escape_identifier();
+fn constructor(
+    escaped_name: &str,
+    access: &str,
+    summary_comment: String,
+    namespace: &str,
+    fields: &[&Field],
+    base_fields: &[&Field],
+) -> CodeBlock {
+    let mut code = CodeBlock::default();
 
-    let namespace = &exception_def.namespace();
+    let mut builder = FunctionBuilder::new(access, "", escaped_name, FunctionType::BlockBody);
 
-    let all_fields = exception_def.all_fields();
-
-    let base_parameters = if let Some(base) = exception_def.base_exception() {
-        base.all_fields().iter().map(|m| m.parameter_name()).collect::<Vec<_>>()
-    } else {
-        vec![]
-    };
-
-    let mut ctor_builder = FunctionBuilder::new("public", "", &exception_name, FunctionType::BlockBody);
-
-    if all_fields.iter().any(|f| f.is_required()) {
-        ctor_builder.add_attribute("global::System.Diagnostics.CodeAnalysis.SetsRequiredMembers");
+    if fields.iter().any(|f| f.is_required()) || base_fields.iter().any(|f| f.is_required()) {
+        builder.add_attribute("global::System.Diagnostics.CodeAnalysis.SetsRequiredMembers");
     }
 
-    ctor_builder.add_comment(
-        "summary",
-        format!(r#"Constructs a new instance of <see cref="{}" />."#, &exception_name),
-    );
+    builder.add_comment("summary", summary_comment);
+
+    builder.add_base_parameters(&base_fields.iter().map(|m| m.parameter_name()).collect::<Vec<String>>());
 
-    for field in &all_fields {
-        ctor_builder.add_parameter(
-            &field.data_type().field_type_string(namespace),
-            field.parameter_name().as_str(),
+    for field in base_fields.iter().chain(fields.iter()) {
+        builder.add_parameter(
+            &field.data_type.field_type_string(namespace),
+            &field.parameter_name(),
             None,
             field.formatted_doc_comment_summary(),
         );
     }
-    ctor_builder.add_base_parameters(&base_parameters);
 
-    // ctor impl
-    let mut ctor_body = CodeBlock::default();
-    for field in exception_def.fields() {
-        let field_name = field.field_name();
-        let parameter_name = field.parameter_name();
+    builder.set_body({
+        let mut code = CodeBlock::default();
+        for field in fields {
+            writeln!(code, "this.{} = {};", field.field_name(), field.parameter_name(),);
+        }
+        code
+    });
 
-        writeln!(ctor_body, "this.{field_name} = {parameter_name};");
-    }
+    code.add_block(builder.build());
+
+    code
+}
+
+fn encode_and_decode(def: &Exception) -> CodeBlock {
+    let mut code = CodeBlock::default();
+
+    let fields = def.fields();
+    let has_base = def.base_exception().is_some();
+
+    let encode_class = FunctionBuilder::new("protected override", "void", "EncodeCore", FunctionType::BlockBody)
+        .add_parameter("ref SliceEncoder", "encoder", None, None)
+        .set_body({
+            let mut code = CodeBlock::default();
```

Comment:
More shadowed `code` variables...
Here the outer code is actually useful, so we shouldn't delete it.
But, we should absolutely change one of their names to avoid shadowing here.

Reply:
I let you fix this in a follow-up PR. Don't forget to update both class_generator and exception_generator.